### PR TITLE
Merging Front end into Main

### LINKED ISSUE #123: Create Recipe Info Web Component
This custom Web Component should have the html tag of `<recipe-info>` and will be rendered with a recipe's JSON using `.data`, similar to the custom web component we made in Lab 6. This `<recipe-info>` component will be placed into a `<recipe-page>` component.

The information from `.data`, (like prepTime, directions, name, rating), can be extracted in a similar manner as to how the custom web components in RecipeCard.js or RecipeExpand.js extract their recipe information from `.data`. You can also look at my code for the RecipeCard.js `<recipe-card>` I created and base it off of that. 

You can assume the `data` passed to this component will be in the same form as the `data` we obtain in the Lab 6 custom component, as well as the `data` parameter received in the `data(data)` function in the RecipeCard.js we have created so far.

This component should only concern the following portion of the figma:
image

The buttons do not need to be there yet, just the information from the recipe for now. 

Feel free to make your own components to be used for this component, e.g. a `<directions-list>` if that would be easier than just making a for loop for `<li>`.

### LINKED ISSUE #124: Create Recipe Page Web Component
This custom Web Component should have the tag of `<recipe-page>` and will be rendered when a `<recipe-card>` is clicked. The corresponding recipe data will be passed into this component using `.data`, similar to the web components previously done in lab 6 and the current version of our RecipeCard.js custom web component. 

The information from .data, (like prepTime, directions, name, rating), can be extracted in a similar manner as to how the custom web components in RecipeCard.js or RecipeExpand.js extract their recipe information from .data. You can also look at my code for the RecipeCard.js <recipe-card> I created and base it off of that.

You can assume the data passed to this component will be in the same form as the data we obtain in the Lab 6 custom component, as well as the data parameter received in the data(data) function in the RecipeCard.js we have created so far.

This component uses the `<recipe-info>` component and should pass the `.data` to the `<recipe-info>` component. #123
This component uses the `<ingredients-info>` component and should pass the `.data` to the `<ingredients-info>` component. #125

This component is concerned with the following portion of the figma:

image

### LINKED ISSUE #125: Create Ingredients Web Component
This custom Web Component should have the html tag of `<ingredients-info>` and will be rendered when a button in `<recipe-info>` in the same `<recipe-page>` is clicked. (You don't have to bind this button yet).

`main.js` will take care of the button bindings. 

The information from .data, (like prepTime, directions, name, rating, etc), can be extracted in a similar manner as to how the custom web components in RecipeCard.js or RecipeExpand.js extract their recipe information from .data. You can also look at my code for the RecipeCard.js I created and base it off of that.

You can assume the data passed to this component will be in the same form as the data we obtain in the Lab 6 custom component, as well as the data parameter received in the data(data) function in the RecipeCard.js we have created so far.

This component is concerned with the following portion of the figma:
image

The quantity scaling and grocery cart additions can be implemented later. The important part is having a list of ingredients.

### LINKED ISSUE #126: Create Nav Bar Component
This component should have the tag of `<nav-bar>` and is always rendered at the top of the page, regardless of whatever page the app is in. The current page will be passed in using `.data`. This will contain the following object:
```
{
  recipeName: {string} containing name of the current recipe being looked at, null if app is not looking at a recipe,
  isRecipe: {Boolean} true if the current page the app is in is a `<recipe-page>`.
}
```

This component is concerned with the following portion of the figma:

image
image

The search box functionality can be implemented later.

### LINKED ISSUE #129: Create Directions-Info Component
This custom web component should have the html tag of `<directions-info>` and will be rendered with a recipe's JSON using `.data` , similar to the custom web component we made in Lab 6. This `<directions-info>` component into a `<recipe-info>` component in issue #123.

The information from `.data`, (like prepTime, directions, name, rating), can be extracted in a similar manner as to how the custom web components in RecipeCard.js or RecipeExpand.js extract their recipe information from .data.

You can assume the data passed to this component will be in the same form as the data we obtain in the Lab 6 custom component, as well as the data parameter received in the data(data) function in the RecipeCard.js we have created so far.

This concerns the following portion of the figma:

image

Feel free to make further components for the list entries if need be.
Checkboxes do not need to be saved for now.

### PR DESCRIPTION
**Pull request by TuckerFrandsen**
## Summary: 
This PR adds/improves upon the HTML/CSS files that attempt to construct the FIGMA diagram of the recipe manager on a single page index.html. It adds several JS components in the form of Recipe Page Component #124, Recipe Info Component #123, Ingredients Component #125, Navigation Bar Component #126, and Directions Component #129. These components recreate the look of the HTML/CSS files that and allow for them to be easily swapped in and out to mimic the feel of switching pages.
 
## index.html
index.html was rewritten to be a mostly blank page in which we would insert our web components through main.js

## Router.js
This module is in charge of storing the functions needed to render the pages and navigating between them. It is heavily based off of Lab 7's router, with minimal tweaking. 

## storage.js
This module is in charge of retrieving/adding data between localStorage and the webapp. It contains the following new functions
- getRecipes()
- isSaved()
- retrieveJSONFromPage()
- hasRecipe()

It modifies the previously defined functions
- addRecipeToSaved()

This module used to manage web components in the previous iteration of the main branch, but that job has since been moved to main.js.

## main.js
This module controls the rendering of each web component and their interactions with each other as well as handling user interaction on the components. It utilizes storage.js in order to process information within the webapp and its components and uses Router.js to navigate between page states.

The major parts implemented are:
- recipesPage()
- recipePage(recipeUrl, recipeJSON)
- renderRecipes(list, target)

## /source/json
This folder is intended to store the recipes we currently are experimenting with before moving onto retrieving recipes through an external API or by scraping.

## Recipe Page Component: 
Calls the Navigation Component, Recipe Info Component, and Directions Component to populate the page based on whatever recipe was clicked on and passed to the function. Fixes #124.

## Recipe Info Component: 
Calls Ingredients Component. Also populates the page with the image of the recipe, ratings, cook times, and adds buttons to show ingredients through Ingredients Component, show nutrition (not yet implemented), and to allow the user to save the recipe (not yet implemented). Fixes #123.

## Ingredients Component: 
Populates a list of ingredients for the user so that they can adjust the serving size and watch the ingredients change (not fully implemented). In addition it also will allow the users to add the ingredients they wish to a grocery list that they can print off and take with them. Fixes #125. 

## Navigation Component: 
Adds a navigation bar that will allow the user to travel between homepage, saved recipes, search bar, etc. Fixes #126. 

## Directions Component: 
Adds the directions for the recipe on the recipe page. Will allow the user to close/expand each step so that they can focus on each one as they go. Fixes #129.

The CSS currently in place for the components is by no means final since the aim was to get functional code working before tidying it up.

Additionally there still needs to be comments added to each file detailing the functionality.

### CODE REVIEW

File: source/js/Directions.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+      }
+    });
+    return value;
+  }
+  
+  /**
+   * Similar to getUrl(), this function extracts the organizations name from the
+   * schema JSON object. It's not in a standard location so this function helps.
+   *
+   * @param {Object} data Raw recipe JSON to find the org string of
```
Comment (TuckerFrandsen) [inline source/js/Directions.js:82]: Codacy has a fix for the issue: [Sentence must end with a period.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
   * @param {Object} data Raw recipe JSON to find the org string of.
```

File: source/js/Directions.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        if (obj['@type'] == 'Recipe') {
+          value = obj['name'];
+        }
+      });
+    }
+    return value;
+  }
+  
+  /**
+   * Extract the URL from the given recipe schema JSON object
```
Comment (TuckerFrandsen) [inline source/js/Directions.js:116]: Codacy has a fix for the issue: [Sentence must end with a period.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
   * Extract the URL from the given recipe schema JSON object.
```

File: source/js/IngredientsInfo.js [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    subtractQuantity.textContent = "-";
+    form.appendChild(subtractQuantity);
+
+    let quantity = document.createElement("p");
+    quantity.classList.add("quantity");
+    const servings = searchForKey(data, "recipeYield");
+    quantity.textContent = servings[0];
+    form.appendChild(quantity);
+
+    let addQuantity = document.createElement("button");
```
Comment (TuckerFrandsen) [inline source/js/IngredientsInfo.js:143]: Codacy has a fix for the issue: ['addQuantity' is never reassigned. Use 'const' instead.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    const addQuantity = document.createElement("button");
```

File: source/js/RecipeExpand.js [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+   * @param {Object} data Raw recipe JSON to find the image of
+   * @returns {String} If found, returns the URL of the image as a string, otherwise null
+   */
+  function getImage(data) {
+    if (data.image?.url) return data.image.url;
+    if (data.image?.contentUrl) return data.image.contentUrl;
+    if (data.image?.thumbnail) return data.image.thumbnail;
+    if (data['@graph']) {
+      for (let i = 0; i < data['@graph'].length; i++) {
+        if (data['@graph'][i]['@type'] == 'ImageObject') {
```
Comment (TuckerFrandsen) [inline source/js/RecipeExpand.js:364]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
      if (data['@graph'][i]['@type'] == 'ImageObject') {
```

File: source/js/RecipeInfo.js [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+          padding: none;
+        }
+
+        button {
+          height: 5vh;
+          width: 10vw;
+        }
+        `;
+
+    const styleElem = document.createElement("style");
```
Comment (TuckerFrandsen) [inline source/js/RecipeInfo.js:61]: Codacy has a fix for the issue: [The variable `styleElem` should be named `styleElement`. A more descriptive name will do too.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    const styleElement = document.createElement("style");
```

File: source/js/GliderRecipe.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        count: tempRating.ratingCount,
+        score: tempRating.ratingValue
+      }
+    }
+
+    // alright cool we got everything, time to actually build it
+
+    // this is gonna be the root element we'll slap everything else onto
+    const entry = document.createElement('li');
+    entry.classList.add('glide__slide');
```
Comment (TuckerFrandsen) [inline source/js/GliderRecipe.js:49]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    entry.classList.add("glide__slide");
```

File: source/js/Router.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+   */
+  navigate(page, statePopped) {
+    // check if we have the route
+    if (!this[page]) {
+      console.log(`${page} does not exist.`);
+      return;
+    }
+
+    let hash = "#" + page;
+    if (page == "home") hash = "";
```
Comment (TuckerFrandsen) [inline source/js/Router.js:39]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    if (page == 'home') {hash = '';}
```

File: source/js/RecipeCard.js [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    }
+    if (!!cleanData.totalTime) {
+      time.textContent = cleanData.totalTime;
+    }
+    info.appendChild(time);
+
+    card.appendChild(info);
+
+    this.shadowRoot.appendChild(styleElem);
+    this.shadowRoot.appendChild(card);
```
Comment (TuckerFrandsen) [inline source/js/RecipeCard.js:107]: Codacy has a fix for the issue: [Prefer `Node#append()` over `Node#appendChild()`.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    this.shadowRoot.append(card);
```

File: source/js/IngredientsInfo.js [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    line.classList.add("line");
+    info.appendChild(line);
+
+    let form = document.createElement("div");
+    form.classList.add("form");
+
+    let addToCart = document.createElement("button");
+    addToCart.classList.add("cartButton");
+    addToCart.textContent = "Add Ingredients To Cart";
+    form.appendChild(addToCart);
```
Comment (TuckerFrandsen) [inline source/js/IngredientsInfo.js:125]: Codacy has a fix for the issue: [Prefer `Node#append()` over `Node#appendChild()`.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    form.append(addToCart);
```

File: source/js/RecipeExpand.js [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+     * Returns the object of the currect recipe being used.
+     */
+    get data() {
+      return this.json;
+    }
+  }
+  
+  /**
+   * Recursively search for a key nested somewhere inside an object
+   * @param {Object} object the object with which you'd like to search
```
Comment (TuckerFrandsen) [inline source/js/RecipeExpand.js:255]: Codacy has a fix for the issue: [There must be a hyphen before @param description.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
   * @param {Object} object - the object with which you'd like to search
```

File: source/js/IngredientsInfo.js [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+ * Extract the URL from the given recipe schema JSON object
+ * @param {Object} data Raw recipe JSON to find the URL of
+ * @returns {String} If found, it returns the URL as a string, otherwise null
+ */
+function getUrl(data) {
+  if (data.url) return data.url;
+  if (data["@graph"]) {
+    for (let i = 0; i < data["@graph"].length; i++) {
+      if (data["@graph"][i]["@type"] == "Article")
+        return data["@graph"][i]["@id"];
```
Comment (TuckerFrandsen) [inline source/js/IngredientsInfo.js:245]: Codacy has a fix for the issue: [Expected no linebreak before this statement.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    }
```

File: source/js/RecipeCard.js [hunk 12]
```diff
@@ -0,0 +1,251 @@
+class RecipeCard extends HTMLElement {
+  constructor() {
+    super();
+    let shadow = this.attachShadow({mode: 'open'});
+  }
+
+  set data(data) {
+    const style=`
```
Comment (TuckerFrandsen) [inline source/js/RecipeCard.js:8]: Codacy has a fix for the issue: [Operator '=' must be spaced.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    const style = `
```

File: source/js/GliderRecipe.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+/**
+ * Takes in a list of ingredients raw from imported data and returns a neatly
+ * formatted comma separated list.
+ * @param {Array} ingredientArr The raw unprocessed array of ingredients from the
+ *                              imported data
+ * @return {String} the string comma separate list of ingredients from the array
+ */
+function createIngredientList(ingredientArr) {
```
Comment (TuckerFrandsen) [inline source/js/GliderRecipe.js:174]: Codacy has a fix for the issue: [Missing space before function parentheses.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
function createIngredientList (ingredientArr) {
```

File: source/js/Directions.js [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        `;
+
+    const styleElem = document.createElement('style');
+    styleElem.innerHTML=style;
+
+    let directionList = searchForKey(data, 'recipeInstructions');
+
+    const wrapper = document.createElement('article');
+    wrapper.classList.add('example');
```
Comment (TuckerFrandsen) [inline source/js/Directions.js:20]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    wrapper.classList.add("example");
```

File: source/js/Directions.js [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+  /**
+   * Extract the URL from the given recipe schema JSON object
+   * @param {Object} data Raw recipe JSON to find the URL of
+   * @returns {String} If found, it returns the URL as a string, otherwise null
+   */
+  function getUrl(data) {
+    if (data.url) return data.url;
+    if (data['@graph']) {
+      for (let i = 0; i < data['@graph'].length; i++) {
+        if (data['@graph'][i]['@type'] == 'Article') return data['@graph'][i]['@id'];
```
Comment (TuckerFrandsen) [inline source/js/Directions.js:124]: Codacy has a fix for the issue: [Expected indentation of 6 spaces but found 8.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
      if (data['@graph'][i]['@type'] == 'Article') return data['@graph'][i]['@id'];
```

File: source/js/RecipeExpand.js [hunk 16]
```diff
@@ -0,0 +1,496 @@
...
+      } else {
+        rating.innerHTML = `
+          <span>No Reviews</span>
+        `;
+      }
+  
+      // Set Ingredients
+      const ingredients = getIngredients(data);
+      ingredients.forEach(ingredient => {
+        const listItem = document.createElement('li');
```
Comment (TuckerFrandsen) [inline source/js/RecipeExpand.js:231]: Codacy has a fix for the issue: [Expected blank line after variable declarations.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
        const listItem = document.createElement('li');
```

File: source/js/RecipeCard.js [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+    card.appendChild(photo);
+
+    const title = document.createElement('p');
+    title.textContent = cleanData.title;
+    card.appendChild(title);
+
+    const info = document.createElement('div');
+    info.classList.add('rating-time');
+
+    const rating = document.createElement('p');
```
Comment (TuckerFrandsen) [inline source/js/RecipeCard.js:71]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    const rating = document.createElement("p");
```

File: source/js/RecipeExpand.js [hunk 18]
```diff
@@ -0,0 +1,496 @@
...
+    if (Array.isArray(categories)) categories = categories.join(', ');
+    return categories.toLowerCase();
+  }
+  
+  /**
+   * Extract the description of the recipe from the given recipe schema JSON obejct
+   * @param {Object} data Raw recipe JSON to find the image of
+   * @returns {String} If found, returns the recipe description
+   */
+  function getDescription(data) {
```
Comment (TuckerFrandsen) [inline source/js/RecipeExpand.js:341]: Codacy has a fix for the issue: [Expected indentation of 0 spaces but found 2.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
function getDescription(data) {
```

File: source/js/RecipeInfo.js [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+/**
+ * Similar to getOrganization(), this extracts recipe name from raw JSON
+ * @param {Object} Data Raw recipe JSON to find name of
+ * @returns {String} if found, returns the name of recipe as string, otherwise null
+ */
+function getRecipeTitle(data) {
+  if (data.name) return data.name;
+  let value = null;
+  if (data["@graph"]) {
+    data["@graph"].forEach((obj) => {
```
Comment (TuckerFrandsen) [inline source/js/RecipeInfo.js:232]: Codacy has a fix for the issue: [There must be a hyphen before @param description.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
 * @param {Object} data - Raw recipe JSON to find the URL of
```

File: source/js/main.js [hunk 20]
```diff
@@ -0,0 +1,299 @@
+import { storage } from "./storage.js";
```
Comment (TuckerFrandsen) [inline source/js/main.js:1]: Codacy has a fix for the issue: [There should be no space after '{'.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
import {storage } from "./storage.js";
```

File: source/js/RecipeExpand.js [hunk 21]
```diff
@@ -0,0 +1,496 @@
...
+    }
+    return null;
+  }
+  
+  /**
+   * Extract the yield of the recipe from the given recipe schema JSON obejct
+   * @param {Object} data Raw recipe JSON to find the image of
+   * @returns {String} If found, returns the recipe yield
+   */
+  function getYield(data) {
```
Comment (TuckerFrandsen) [inline source/js/RecipeExpand.js:296]: Codacy has a fix for the issue: [Expected indentation of 0 spaces but found 2.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
function getYield(data) {
```

File: source/js/Directions.js [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    set data(data) {
+        const style=`
+        .example {
+
+        }
+        `;
+
+    const styleElem = document.createElement('style');
```
Comment (TuckerFrandsen) [inline source/js/Directions.js:14]: Codacy has a fix for the issue: [Expected blank line after variable declarations.](https://app.codacy.com/gh/cse110-fa21-group27/cse110-fa21-group27/pullRequest?prid=8447291)
```suggestion
    const styleElem = document.createElement('style');
```

**Review by mlgi — APPROVED** [review]
Components seem to be good, we're definitely going to need to update them later. Back-end has a handful of bugs and more features needed, so we'll go ahead and create issues for them for the next sprint! We also still need to make it behave look like the design team's figma. Good demo for first sprint!

**Review by Steven-Christensen — APPROVED** [review]
thanks for putting this all together!

**Pull request merged**